witness/src/Dscp/Witness/Web/Logic.hs

```diff
@@ -209,19 +209,27 @@ getHashType someHash = fmap (fromMaybe HashIsUnknown) . runMaybeT . asum $
 -- | Check @'FairCV'@ against records in the public chain.
 checkFairCV :: forall ctx m. WitnessWorkMode ctx m => FairCV -> m FairCVCheckResult
 checkFairCV =
-    fmap buildResults . checkAgainstDB . readyFairCV
+    fmap buildResults . checkCV . readyFairCV
   where
-    checkAgainstDB (FairCVReady cv) =
-        M.traverseWithKey (M.traverseWithKey . checkProofAgainstDB) cv
+    checkCV (FairCV sAddr _ cv) =
+        M.traverseWithKey (M.traverseWithKey . checkProofAgainstDB sAddr) cv
 
-    checkProofAgainstDB addr h proof =
-        maybe False (checkProofPure addr proof) <$>
+    checkProofAgainstDB sAddr eAddr h proof =
+        maybe False (checkProofPure sAddr eAddr proof) <$>
             runSdMempoolLocked (getPublicationByHeaderHash h)
 
-    checkProofPure addr proof ptw =
+    checkProofPure sAddr eAddr proof ptw =
         let root = ptw ^. ptwTxL.ptHeaderL.pbhBodyProof
-        in verifyPubTxWitnessed addr ptw &&
-           root == mprRoot proof
+            pubAuthor = ptw ^. ptwTxL.ptAuthorL
+        in verifyPubTxWitnessed ptw &&
+           eAddr == pubAuthor &&
+           root == mprRoot proof &&
+           all (checkTxSubmission sAddr) (mprProof proof)
+
+    checkTxSubmission sAddr (PrivateTx sSub _ _) =
+        let sAuthor = sSub ^. ssSubmission.sStudentId
+        in sAuthor == sAddr && isRight (validateSubmission sSub)
```

Comment:
Really, we should check all submissions since educator could accept an invalid submission.
(Or, can't we actually verify submissions once they appear in public chain - that would automatically mean that any transaction witness knows about here is valid on itself? :thinking:)

Anyway, that's a bit unpleasant that we have this little logic duplicated - one another place is implementation of Student API, and `verifyStudentSubmission` function seems to do exactly the same which `checkTxSubmission` does. Could we unite them somehow?

Reply:
Submissions never appear in the public chain - they are inhabitants of private chains and FairCVs. So, when Witness validates FairCV, all the submissions there would be most likely witnessed for the first time.

I can move `verifyStudentSubmission` to `Dscp.Core.Validation.Educator` too, why not.

Reply:
Ah right, I forgot that we keep tree without nodes in public blocks.